aten/src/ATen/native/native_functions.yaml

```diff
@@ -4590,13 +4590,13 @@
 - func: _weight_norm_differentiable_backward(Tensor grad_w, Tensor saved_v, Tensor saved_g, Tensor saved_norms, int dim) -> (Tensor, Tensor)
   variants: function
 
-- func: zeros.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+- func: zeros.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
   device_check: NoCheck
   device_guard: False
 
-- func: zeros(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+- func: zeros(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
 
-- func: zeros.out(int[] size, *, Tensor(a!) out) -> Tensor(a!)
+- func: zeros.out(int[] size, *, MemoryFormat? memory_format=None, Tensor(a!) out) -> Tensor(a!)
```

Comment:
This is bc-incompatible? cc @mruberry

Reply:
Is that because the new optional arg isn't at the end?

I'm not sure that we want to add the argument to `zeros.out` anyway. It doesn't take in any other TensorOptions params

Reply:
It looks like the codegen as-is requires us to add memory_format here support to `torch.zeros` :(

The codegen makes certain assertions about every out= operator having a corresponding out-of-place variant with the same schema.

Although I'm wondering - it looks like the python binding codegen allows TensorOptions args to be different between the out= and out-of-place variants - do we want to treat memory_format the same way?

I guess it just comes down to whether or not we want out= factory functions to take in a memory_format arguments. They already don't take in TensorOptions arguments.

Reply:
This should work OK because it's an optional non-list kwarg added at the end of the argument list or just before the out kwarg.

Reply:
Ok, good to know that adding it won't be BC-breaking

@ezyang regardless, do you think it makes sense to add a `MemoryFormat` argument to the `out` variant of a factory function? They current don't take in any `TensorOptions` arguments, because those just get inferred from the passed-in`out` tensor. I could imagine wanting to do the same thing with MemoryFormat (inferring from the out tensor instead of explicitly passing it in), but it'll require some changes to the python codegen.

aside: out= factory functions also seem kinda weird in general to me. We're not treating the out tensor as an opaque buffer, since we're inferring some metadata from it (like the strides, etc). But `zeros.out` also accepts a `sizes` argument, allowing you to resize the `out` tensor (instead of, say, not accepting an additional arguments and just setting the passed-in-tensor data to be all 0's).

Reply:
I don't know why we have an out variant of zeros. Just use `zero_` for crying out loud.

If skipping the out overload makes this patch simpler I'm all for it. I don't have all of the complications paged in though right now.